packages/aws-cdk/lib/api/garbage-collection/garbage-collector.ts

```diff
@@ -0,0 +1,389 @@
+      // where gc is run for the first time on a long-standing bucket where ~100% of objects are isolated.
+      for await (const batch of this.readBucketInBatches(s3, bucket, batchSize, currentTime)) {
+        print(chalk.green(`Processing batch ${batches} of ${Math.floor(numObjects / batchSize) + 1}`));
+        printer.start();
+
+        const { isolated, notIsolated } = batch.reduce<{ isolated: S3Asset[]; notIsolated: S3Asset[] }>((acc, obj) => {
+          if (!activeAssets.contains(obj.fileName())) {
+            acc.isolated.push(obj);
+          } else {
+            acc.notIsolated.push(obj);
+          }
+          return acc;
+        }, { isolated: [], notIsolated: [] });
+
+        debug(`${isolated.length} isolated assets`);
+
+        let deletables: S3Asset[] = isolated;
+        let taggables: S3Asset[] = [];
+
+        if (graceDays > 0) {
+          debug('Filtering out assets that are not old enough to delete');
+          await this.parallelReadAllTags(s3, isolated);
+          // eslint-disable-next-line @cdklabs/promiseall-no-unbounded-parallelism
+          deletables = await Promise.all(isolated.map(async (obj) => {
+            const shouldDelete = await obj.isolatedTagBefore(s3, new Date(currentTime - (graceDays * DAY)));
+            return shouldDelete ? obj : null;
+          })).then(results => results.filter((obj): obj is S3Asset => obj !== null));
+
+          // eslint-disable-next-line @cdklabs/promiseall-no-unbounded-parallelism
+          taggables = await Promise.all(isolated.map(async (obj) => {
+            const shouldTag = await obj.noIsolatedTag(s3);
+            return shouldTag ? obj : null;
+          })).then(results => results.filter((obj): obj is S3Asset => obj !== null));
+        }
+
+        debug(`${deletables.length} deletable assets`);
+        debug(`${taggables.length} taggable assets`);
+
+        if (this.permissionToDelete && deletables.length > 0) {
+          await this.parallelDelete(s3, bucket, deletables, printer);
+        }
+
+        if (this.permissionToTag && taggables.length > 0) {
+          await this.parallelTag(s3, bucket, taggables, currentTime, printer);
+        }
+
+        printer.reportScannedObjects(batch.length);
+
+        // We untag objects that are referenced in ActiveAssets and currently have the Isolated Tag.
+        let untaggables: S3Asset[] = await Promise.all(notIsolated.map(async (obj) => {
+          const noTag = await obj.noIsolatedTag(s3);
+          return noTag ? null : obj;
+        })).then(results => results.filter((obj): obj is S3Asset => obj !== null));
+
+        debug(`${untaggables.length} assets to untag`);
+
+        if (this.permissionToTag && untaggables.length > 0) {
+          await this.parallelUntag(s3, bucket, untaggables);
+        }
+      }
+    } catch (err: any) {
+      throw new Error(err);
+    } finally {
+      backgroundStackRefresh.stop();
+      printer.stop();
+      clearTimeout(timeout);
```

Comment:
This should not be necessary, the backgroundrefresh should be managing this.

Reply:
not sure how backgroundStackRefresh takes care of this -- where is `.stop()` supposed to be called if not here?